src/PromiseInterface.php

```diff
@@ -28,9 +31,9 @@ interface PromiseInterface
      *  2. `$onFulfilled` and `$onRejected` will never be called more
      *      than once.
      *
-     * @param callable|null $onFulfilled
-     * @param callable|null $onRejected
-     * @return PromiseInterface
+     * @template TFulfilled as PromiseInterface<T>|T
+     * @param (callable(T): TFulfilled)|null $onFulfilled
+     * @return PromiseInterface<T>
```

Comment:
Shouldn't this be `@return PromiseInterface<TFulfilled>` (plus additional type checks if the function returns `PromiseInterface<TFulfilled>`).

Should probably add some tests similar to this:

```php
assertType('React\Promise\PromiseInterface<bool>', resolve(true));

assertType('React\Promise\PromiseInterface<int>', resolve(true)->then(fn(bool $value): int => (int) $value));

assertType('React\Promise\PromiseInterface<int>', resolve(true)->then(fn(bool $value): PromiseInterface => resolve((int) $value)));
```

Not sure how much of this is supported, WDYT?

Reply:
This would be really good to have, it is also a really hard one. Here is the PHPStan playground with my work on it so far: https://phpstan.org/r/bf62f95b-4560-4115-9eac-15518640f137

Shifting my focus to your other comments for the rest of today before coming back to this one.

Reply:
Resolved :shipit: